### PR TITLE
Radial BNN (ResNet50) for Diabetic Retinopathy

### PR DESCRIPTION
**Pull request by nband**

### CODE REVIEW

File: uncertainty_baselines/models/radial_utils.py [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+
+    # Kl = Entropy - Cross-entropy
+    # We find cross-entropy by MC estimation
+    n_samples = 10
+    cross_entropy = tf.math.reduce_sum(
+      q.log_prob(p.sample((n_samples,)))) / n_samples
+    return p.entropy() - cross_entropy
+
+
+@kullback_leibler.RegisterKL(Radial, Radial)
```
Comment (SebFar) [inline uncertainty_baselines/models/radial_utils.py:271]: I've recently learned that I made a mistake in equation (43) which means that this expression is missing a term. I'm going to derive the correction term but haven't done it yet. For now maybe leave this KL out?

File: baselines/diabetic_retinopathy_detection/radial.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+  'kl_annealing_epochs', 200,
+  'Number of epochs over which to anneal the KL term to 1.')
+flags.DEFINE_float('prior_stddev', 0.1, 'Fixed stddev for weight prior.')
+flags.DEFINE_float(
+  'stddev_init', 1e-3,
+  'Initialization of posterior standard deviation parameters.')
+
+# General model flags.
+flags.DEFINE_integer('seed', 42, 'Random seed.')
+flags.DEFINE_float('l2', 5e-5, 'L2 regularization coefficient.')
```
Comment (SebFar) [inline baselines/diabetic_retinopathy_detection/radial.py:90]: Why use an L2 reg? The ELBO ought to serve that purpose, at least if you're not doing the tying to the mean thing. I'm very confused why one would compute the KL to a tied prior and then separately regularize the mean values back to zero?

File: baselines/diabetic_retinopathy_detection/radial.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+        decay_ratio=FLAGS.lr_decay_ratio,
+        decay_epochs=lr_decay_epochs,
+        warmup_epochs=FLAGS.lr_warmup_epochs)
+    optimizer = tf.keras.optimizers.SGD(
+      lr_schedule, momentum=0.9, nesterov=True)
+    metrics = utils.get_diabetic_retinopathy_base_metrics(
+      use_tpu=use_tpu, num_bins=FLAGS.num_bins)
+    metrics.update({
+      'train/kl': tf.keras.metrics.Mean(),
+      'train/kl_scale': tf.keras.metrics.Mean()})
```
Comment (SebFar) [inline baselines/diabetic_retinopathy_detection/radial.py:186]: Are these supposed to b ethe same?

File: baselines/diabetic_retinopathy_detection/radial.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+      checkpoint.restore(latest_checkpoint)
+      logging.info('Loaded checkpoint %s', latest_checkpoint)
+      initial_epoch = optimizer.iterations.numpy() // steps_per_epoch
+
+  # Finally, define OOD metrics outside the accelerator scope for CPU eval.
+  # This will cause an error on TPU.
+  if not use_tpu:
+    metrics.update({
+      'train/auc': tf.keras.metrics.AUC(),
+      'test/auc': tf.keras.metrics.AUC()})
```
Comment (SebFar) [inline baselines/diabetic_retinopathy_detection/radial.py:202]: Also supposed to be same?

File: baselines/diabetic_retinopathy_detection/radial.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+
+        l2_loss = FLAGS.l2 * 2 * tf.nn.l2_loss(
+            tf.concat(filtered_variables, axis=0))
+        kl = sum(model.losses)
+        kl_scale = tf.cast(optimizer.iterations + 1, kl.dtype)
+        kl_scale /= steps_per_epoch * FLAGS.kl_annealing_epochs
+        kl_scale = tf.minimum(1., kl_scale)
+        kl_loss = kl_scale * kl
+
+        loss = negative_log_likelihood + l2_loss + kl_loss
```
Comment (SebFar) [inline baselines/diabetic_retinopathy_detection/radial.py:239]: I really don't think this should use an L2 loss. Definitely if you aren't using tied means it should not have l2 loss.

Comment (SebFar) [inline baselines/diabetic_retinopathy_detection/radial.py:239]: I'm not sure these are scaled correctly? The NLL term is averaged per example? But the KL is the complete model KL. It will have to be divided by the number of examples in the training epoch in order to be in the same units?

File: uncertainty_baselines/models/resnet50_radial.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+  bn_name_base = 'bn' + str(stage) + block + '_branch'
+  kernel_regularizer = get_radial_kernel_regularizer(
+    tied_mean_prior=tied_mean_prior, radial_prior=radial_prior)
+
+  x = Conv2DFlipout(
+    filters1,
+    kernel_size=1,
+    kernel_initializer=TrainableHeRadial(
+      stddev_initializer=tf.keras.initializers.TruncatedNormal(
+        mean=np.log(np.expm1(stddev_init)), stddev=0.1)),
```
Comment (SebFar) [inline uncertainty_baselines/models/resnet50_radial.py:111]: Should 0.1 be hard coded here?

File: baselines/diabetic_retinopathy_detection/radial.py (hunk 5 above)
Comment (SebFar) [reply baselines/diabetic_retinopathy_detection/radial.py:239]: Ah, I think I take this back and this is implemented in the regularizer call

**Review by SebFar — COMMENTED** [review]
This looks great, Neil, thanks a ton for doing this. I haven't stepped through the log_prob calculations explicitly yet, which I should probably do to be absolutely sure. Important points to note are:
- My error on the Radial-Radial KL, which needs to be left out until I fix it.
- I'm certain that L2 should not be there unless the means are tied (and I think adding L2 and tieing the means is literally the same as not tieing the means for MVG posterior/prior and is incorrect for Radial posterior or prior. But I'm curious what the original motivation was for this tweak).

File: uncertainty_baselines/models/radial_utils.py [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+      seed=seed,
+      **kwargs)
+
+  def get_config(self):
+    return {
+      'seed': self.seed,
+    }
+
+
+class RadialKLDivergenceWithTiedMean(tf.keras.regularizers.Regularizer):
```
[the remark this answers is not in the document]
Comment (nband) [reply uncertainty_baselines/models/radial_utils.py:392]: Agreed - I added this for comparison to the VI implementation, which has support for a MVG prior with tied mean. 
For the time being (while we don't have a Radial-Radial KL) I'll leave these Radial regularizers (means tied and otherwise) out.

File: uncertainty_baselines/models/radial_utils.py [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+  """KL with Radial prior whose mean is fixed at the variational posterior's."""
```
[the remark this answers is not in the document]
Comment (nband) [reply uncertainty_baselines/models/radial_utils.py:393]: Sounds good, I have that option available! It is already implemented in Edward2 (ed.regularizers.NormalKLDivergence[WithTiedMean])

File: baselines/diabetic_retinopathy_detection/radial.py (hunk 3 above)
Comment (nband) [reply baselines/diabetic_retinopathy_detection/radial.py:186]: Yep, "Mean" just describes the aggregation applied when the metric is computed (we provide the KL and the KL scale respectively)

File: baselines/diabetic_retinopathy_detection/radial.py (hunk 4 above)
Comment (nband) [reply baselines/diabetic_retinopathy_detection/radial.py:202]: Same as above, yes

File: baselines/diabetic_retinopathy_detection/radial.py [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+
+# Radial BNN flags.
+flags.DEFINE_bool(
+  'tied_mean_prior', True,
+  'If True, fix the mean of the prior to that of the variational posterior. '
+  'This causes the KL to only penalize the standard deviation of the weight '
+  'posterior, and not its mean.')
+flags.DEFINE_bool(
+  'radial_prior', False,
+  'If True, use a Radial prior instead of a Normal. This is by default False '
```
[the remark this answers is not in the document]
Comment (nband) [reply baselines/diabetic_retinopathy_detection/radial.py:77]: If radial prior is False we use a MVG prior, which is set in `resnet50_radial.py`.

File: baselines/diabetic_retinopathy_detection/radial.py [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+
+This script performs variational inference with a Radial posterior,
+as introduced in Radial Bayesian Neural Networks (Farquhar et al. 2020).
+
+We use a few additional, noteworthy techniques in training:
+
+1. Normal (or Radial) prior whose mean is tied at the variational posterior's.
+    This makes the KL penalty only penalize the weight posterior's standard
+    deviation and not its mean.
+    The prior's standard deviation is a fixed hyperparameter.
```
[the remark this answers is not in the document]
Comment (nband) [reply baselines/diabetic_retinopathy_detection/radial.py:26]: I agree, and will have the scale set to He's (sqrt(2 / fan_in)) by default!

File: uncertainty_baselines/models/resnet50_radial.py (hunk 6 above)
Comment (nband) [reply uncertainty_baselines/models/resnet50_radial.py:111]: Probably ought to be a hyper, good point!

File: uncertainty_baselines/models/radial_utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (SebFar) [reply uncertainty_baselines/models/radial_utils.py:393]: I'm not sure that will correctly calculate the KL between a Radial and Normal prior? Isn't that designed for Normal to Normal?

File: uncertainty_baselines/models/radial_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (SebFar) [reply uncertainty_baselines/models/radial_utils.py:392]: Makes sense.

File: baselines/diabetic_retinopathy_detection/radial.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (SebFar) [reply baselines/diabetic_retinopathy_detection/radial.py:26]: Great

File: baselines/diabetic_retinopathy_detection/radial.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (SebFar) [reply baselines/diabetic_retinopathy_detection/radial.py:77]: Right, I think what I'd been confused about what that I hadn't seen NormalKLDivergenceWithTiedMean defined, and I think that we'll need to be careful to write it out correctly for Radial-Normal, where the existing thing in Edward will be Normal-Normal? You can get the correct value from the same method as the standard Radial-Normal one just setting the mean value correctly.

File: uncertainty_baselines/models/radial_utils.py [hunk 11]
```diff
@@ -0,0 +1,447 @@
...
+    elif isinstance(other, Radial):
+      return kl_radial_radial(p=self, q=other)
+    else:
+      raise NotImplementedError(f'Other distribution has type {type(other)}.')
+
+
+@kullback_leibler.RegisterKL(
+  Radial, distributions.MultivariateNormalLinearOperator)
+def kl_radial_normal(p, q, name=None):
+  with tf.name_scope(name or 'kl_radial_normal'):
```
Comment (SebFar) [inline uncertainty_baselines/models/radial_utils.py:255]: Should we check that p and q are correct distributions here? Or is that handled elsewhere?

File: baselines/diabetic_retinopathy_detection/radial.py (hunk 5 above)
Comment (nband) [reply baselines/diabetic_retinopathy_detection/radial.py:239]: I agree this would no longer be principled, raising and explaining this now.

File: uncertainty_baselines/models/radial_utils.py (hunk 11 above)
Comment (nband) [reply uncertainty_baselines/models/radial_utils.py:255]: The registry guarantees this, and in normal execution it is checked in Radial.kl_divergence

File: uncertainty_baselines/models/resnet50_radial.py [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+      variational posterior, which causes the KL to only penalize the weight
+      posterior's standard deviation, and not its mean.
+
+  Returns:
+    tf.Tensor.
+  """
+  filters1, filters2, filters3 = filters
+  conv_name_base = 'res' + str(stage) + block + '_branch'
+  bn_name_base = 'bn' + str(stage) + block + '_branch'
+  kernel_regularizer_class = get_kernel_regularizer_class(
```
Comment (SebFar) [inline uncertainty_baselines/models/resnet50_radial.py:78]: I'm not certain, but I think this is calling the standard variational_utils get_kernel_regularizer class which is using whether tied_mean_prior is true to get one of two things that are designed for Normal posterior approximations. Are you sure these are correct here?

Comment (nband) [reply uncertainty_baselines/models/resnet50_radial.py:78]: Yes -- the Ed regularizers end up calling the correct KL between Radial posterior and MVG prior

**Review by SebFar — APPROVED** [review]
Looks good to me, thanks for all the edits!

**Review by znado — APPROVED** [review]
